// @livepeer/core-web/dist/index.js
// src/index.ts
import {
  ACCESS_CONTROL_ERROR_MESSAGE,
  BFRAMES_ERROR_MESSAGE,
  NOT_ACCEPTABLE_ERROR_MESSAGE,
  PERMISSIONS_ERROR_MESSAGE,
  STREAM_OFFLINE_ERROR_MESSAGE,
  STREAM_OPEN_ERROR_MESSAGE,
  isAccessControlError,
  isBframesError,
  isNotAcceptableError,
  isPermissionsError,
  isStreamOfflineError
} from "@livepeer/core/errors";
import {
  createControllerStore,
  getMediaSourceType
} from "@livepeer/core/media";
import {
  createStorage,
  noopStorage
} from "@livepeer/core/storage";
import {
  b64Decode,
  b64Encode,
  b64UrlDecode,
  b64UrlEncode,
  deepMerge,
  omit,
  pick
} from "@livepeer/core/utils";
import { version } from "@livepeer/core/version";
export {
  ACCESS_CONTROL_ERROR_MESSAGE,
  BFRAMES_ERROR_MESSAGE,
  NOT_ACCEPTABLE_ERROR_MESSAGE,
  PERMISSIONS_ERROR_MESSAGE,
  STREAM_OFFLINE_ERROR_MESSAGE,
  STREAM_OPEN_ERROR_MESSAGE,
  b64Decode,
  b64Encode,
  b64UrlDecode,
  b64UrlEncode,
  createControllerStore,
  createStorage,
  deepMerge,
  getMediaSourceType,
  isAccessControlError,
  isBframesError,
  isNotAcceptableError,
  isPermissionsError,
  isStreamOfflineError,
  noopStorage,
  omit,
  pick,
  version
};
